File: cpp/739. [M] Daily Temperatures.cpp
// https://leetcode.com/problems/two-sum/

#include <vector>
#include <unordered_map>
#include <assert.h>
#include <iostream>
#include <stack>

using namespace std;;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int size = temperatures.size();
        vector<int> res(size, 0);
        stack<int> s;
        for (int i = 0; i < size; i++) {
            while (!s.empty() && temperatures.at(s.top()) < temperatures[i]) {
                res.at(s.top()) = i - s.top();
                s.pop();
            }
            s.push(i);
        }
        return res;
    }
};

int main() {
    Solution sol = Solution();
    vector<vector<int>> v = {{1,2},{2,3},{3,4}};
}

File: python/739. [M] Daily Temperatures.py
from typing import List


class Solution:

    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        ans = [0] * n
        stack = []
        for i, temp in enumerate(reversed(temperatures)):
            i = n - i - 1
            while stack and stack[-1][1] <= temp:
                stack.pop()
            if stack:
                ans[i] = stack[-1][0] - i
            stack.append((i, temp))
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.dailyTemperatures([1, 2, 3, 2, 4, 2, 1, 2, 1, 2, 6]))
